content/blog/2025-12-16-gvn-final-report.md

```diff
@@ -0,0 +1,1350 @@
+|--------|---------------|---------------|--------|
+| Static Ratio (GM) | 0.8474x | **0.8397x** | +0.9% better |
+| Dynamic Ratio (GM) | 0.8165x | **0.8142x** | +0.3% better |
+| Benchmarks worse (both) | 30/114 | **22/114** | -27% |
+| Benchmarks improved (static) | 47/114 | **53/114** | +13% |
+
+These optimizations further reduced the number of regressions while maintaining 100% correctness on all 114 benchmarks.
+
+---
+
+## 3. Hardest Parts to Get Right
+
+### 3.1 SSA Overhead: The 162% Bloat Problem
+
+**The Challenge**: Our initial implementation caused the optimizer to **increase** code size by up to 162% on some benchmarks. This was unacceptable — an optimizer should never make code worse.
+
+**Root Cause Analysis**: We spent significant time tracing through the pipeline to understand why. The culprit was minimal SSA placing 73 phi nodes in `fizz-buzz.bril` (a 57-instruction program).
+
+**The Solution**: Implementing pruned SSA with liveness analysis. This required:
+1. Implementing a full backward dataflow analysis for liveness
+2. Modifying phi placement to check liveness
+3. Carefully handling the interaction between liveness and SSA renaming
+
+**Key Insight**: The difference between minimal and pruned SSA is often glossed over in textbooks, but it's critical for practical compilers. Dead variables at join points don't need phi nodes.
+
+### 3.2 Parallel Copy Sequentialization Infinite Loop
+
+**The Challenge**: The SSA lowering pass would hang indefinitely on certain programs.
+
+**Root Cause**: The algorithm for converting parallel copies to sequential copies had a bug in cycle detection. When a destination was used as a source by another copy, the algorithm couldn't make progress.
+
+**The Solution**: Complete rewrite of `_parallel_copies_to_seq()` with proper cycle detection:
+1. Identify safe copies (source won't be overwritten, dest not needed)
+2. If no safe copy, find a cycle variable (both source and dest)
+3. Break the cycle by saving to a temporary
+4. Update remaining copies to use the temporary
+
+**Key Insight**: Parallel copy sequentialization is a classic problem, but subtle to get right. The solution is essentially graph-based: find a topological order if acyclic, break cycles with temps otherwise.
+
+### 3.3 Copy Propagation Soundness
+
+**The Challenge**: Aggressive copy propagation can be unsound if done after SSA lowering, because copies inserted by phi lowering are **necessary** for correctness.
+
+**Example of the bug**:
+```
+# After phi lowering:
+# End of B1:
+x = a    # Copy for phi
+jmp B3
+
+# B3:
+use(x)
+```
+
+If we propagate `x → a` and eliminate the copy, the program breaks because `a` might not be available in B3 (it came from a different predecessor).
+
+**The Solution**: Perform copy propagation **while still in SSA form**, before phi lowering. In SSA, every definition dominates all its uses, so copy propagation is always safe.
+
+### 3.4 Entry Block Creation for Label-First Functions
```

Comment:
This is a pretty good example: this seems like an extremely small change. Was this really one of the _hardest things you encountered_?